### PR TITLE
Configure database and setup authentication

### PR DESCRIPTION
**Pull request by arnoldkunihira**

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -2,3 +2,9 @@ PORT=3333
 HOST=0.0.0.0
 NODE_ENV=development
 APP_KEY=0ttwaajIw4mEJqZx3xLyGjqN1Lj_ur7y
```
Comment (vongohren) [inline .env.example:4]: What is app key :D

File: ace-manifest.json [hunk 2]
```diff
@@ -48,6 +48,219 @@
       "args": [],
       "aliases": [],
       "flags": []
+    },
+    "db:seed": {
```
Comment (vongohren) [inline ace-manifest.json:52]: Interessting, got any docs for this manifest thing?

File: app/Middleware/Auth.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
+import { GuardsList } from '@ioc:Adonis/Addons/Auth'
```
Comment (vongohren) [inline app/Middleware/Auth.ts:1]: How much of this file comes by generation or is written by you?

File: app/Models/User.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    @column()
+    public rememberMeToken?: string;
+
+    @column.dateTime({ autoCreate: true })
+    public createdAt: DateTime;
+
+    @column.dateTime({ autoCreate: true, autoUpdate: true })
+    public updatedAt: DateTime;
+
+    @beforeSave()
```
Comment (vongohren) [inline app/Models/User.ts:27]: ADonis sure gives you everything!
Have you used it alot before?

File: config/auth.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    | OAT (Opaque access tokens) guard uses database backed tokens to authenticate
+    | HTTP request. This guard DOES NOT rely on sessions or cookies and uses
+    | Authorization header value for authentication.
+    |
+    | Use this guard to authenticate mobile apps or web clients that cannot rely
+    | on cookies/sessions.
+    |
+    */
+    api: {
+      driver: 'oat',
```
Comment (vongohren) [inline config/auth.ts:36]: Any special thoughts on this driver?

File: config/database.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
+/**
+ * Config source: https://git.io/JesV9
+ *
+ * Feel free to let us know via PR, if you find something broken in this config
+ * file.
+ */
+
+import Env from '@ioc:Adonis/Core/Env';
+import { DatabaseConfig } from '@ioc:Adonis/Lucid/Database';
```
Comment (vongohren) [inline config/database.ts:9]: Is this just Typescript or ?

File: config/database.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
+/**
+ * Config source: https://git.io/JesV9
+ *
+ * Feel free to let us know via PR, if you find something broken in this config
+ * file.
+ */
+
+import Env from '@ioc:Adonis/Core/Env';
```
Comment (vongohren) [inline config/database.ts:8]: Sure give you everetyhing huh :P Instead of using dotenv

File: contracts/auth.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
+/**
+ * Contract source: https://git.io/JOdz5
+ *
+ * Feel free to let us know via PR, if you find something broken in this
+ * file.
+ */
+
+import User from 'App/Models/User';
+
+declare module '@ioc:Adonis/Addons/Auth' {
```
Comment (vongohren) [inline contracts/auth.ts:10]: What is a contract?

File: database/migrations/1628079946701_users.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
+import BaseSchema from '@ioc:Adonis/Lucid/Schema';
```
Comment (vongohren) [inline database/migrations/1628079946701_users.ts:1]: You run these migrations via adonis?

File: tsconfig.json [hunk 10]
```diff
@@ -1,17 +1,35 @@
...
   "compilerOptions": {
     "outDir": "build",
     "rootDir": "./",
     "sourceMap": true,
     "paths": {
-      "App/*": ["./app/*"],
-      "Config/*": ["./config/*"],
-      "Contracts/*": ["./contracts/*"],
-      "Database/*": ["./database/*"]
+      "App/*": [
```
Comment (vongohren) [inline tsconfig.json:15]: Any reason you want to use capitalized letters on this?

File: ace-manifest.json (hunk 2 above)
Comment (arnoldkunihira) [reply ace-manifest.json:52]: https://docs.adonisjs.com/guides/ace-commandline#generating-the-ace-manifest-file

File: app/Middleware/Auth.ts (hunk 3 above)
Comment (arnoldkunihira) [reply app/Middleware/Auth.ts:1]: The support for auth is added by using **node ace configure @adonisjs/auth** and this command generates for me the files in the image below:
image

Comment (arnoldkunihira) [reply app/Middleware/Auth.ts:1]: Then I get to modify the files I need when the need arises such as the users' migration and its model to suit my needs.

File: app/Models/User.ts (hunk 4 above)
Comment (arnoldkunihira) [reply app/Models/User.ts:27]: Yes, I have used it before from version 4.0 to 4.1 and now on version 5 to create most of the APIs that require the usage of nodejs and typescript in the backend.

File: config/database.ts (hunk 7 above)
Comment (arnoldkunihira) [reply config/database.ts:8]: In a fresh adonisjs installation, the root directory of the application contains a **.env** file that defines common environment variables and this file can not be tracked by version control, so the **import Env from '@ioc:Adonis/Core/Env'** calls the **.env** that is in the root directory with environment variables. 
Now when working in a team, they can not see the **.env** file instead they see the **.env.example** file which contains placeholder values for environment variables. so all they do is type this command **cp .env.example .env** on their terminals and configure **.env** locally on their computers.

File: contracts/auth.ts (hunk 8 above)
Comment (arnoldkunihira) [reply contracts/auth.ts:10]: A contract is basically a way that describes exactly how your program should work within the framework and then checks those contracts such as auth.ts, events.ts at runtime. If something goes wrong, it pinpoints exactly what section of code was to blame with great descriptive error messages.

File: database/migrations/1628079946701_users.ts (hunk 9 above)
Comment (arnoldkunihira) [reply database/migrations/1628079946701_users.ts:1]: yes, I do using **node ace migration:run** command.

File: tsconfig.json (hunk 10 above)
Comment (arnoldkunihira) [reply tsconfig.json:15]: Those are default settings for adonisjs paths for the project when one does a fresh installation and in most cases, the tsconfig.json file is untouched throughout the development process.

File: config/database.ts (hunk 6 above)
Comment (arnoldkunihira) [reply config/database.ts:9]: Yes, that is Typescript

File: .env.example (hunk 1 above)
Comment (arnoldkunihira) [reply .env.example:4]: The APP_KEY is a random 32-character string stored in the APP_KEY in your projects **.env** file. 
It is used for encrypting cookies and creating the signature for signed URLs and queued closures.

File: config/auth.ts (hunk 5 above)
Comment (arnoldkunihira) [reply config/auth.ts:36]: This driver uses database-backed tokens to authenticate HTTP requests and also I can use this driver to authenticate mobile apps or web clients that do not rely on cookies/sessions.

File: .env.example (hunk 1 above)
Comment (vongohren) [reply .env.example:4]: Encryption of cookies, signes URLs and queued closures, is this Adonis out of the box stuff? Or how would you use it?

File: app/Middleware/Auth.ts (hunk 3 above)
Comment (vongohren) [reply app/Middleware/Auth.ts:1]: Interessting! Nice stuff!

File: app/Models/User.ts (hunk 4 above)
Comment (vongohren) [reply app/Models/User.ts:27]: Cool, what is the best key takeaway in your mind?

File: config/auth.ts (hunk 5 above)
Comment (vongohren) [reply config/auth.ts:36]: So any difference on this versus JWTs? Or you got any thoughts on that?

File: config/database.ts (hunk 6 above)
Comment (vongohren) [reply config/database.ts:9]: 👍

File: config/database.ts (hunk 7 above)
Comment (vongohren) [reply config/database.ts:8]: Yeah I get you :D npm dotenv package is the same. Maybe what they use behind this package.
We use a combination of .env, and injecting during runtime

File: contracts/auth.ts (hunk 8 above)
Comment (vongohren) [reply contracts/auth.ts:10]: Interessting! Again, what is your feeling of benefits over any other webframework? Express or what not?

File: tsconfig.json (hunk 10 above)
Comment (vongohren) [reply tsconfig.json:15]: Okoko!

File: .env.example (hunk 1 above)
Comment (arnoldkunihira) [reply .env.example:4]: This is Adonis out-of-the-box stuff, however, the app_key is a private string that nobody knows about, so if only your application knows the app_key, only your application can decrypt the data that is encrypted by this key, which is a security advantage and that is why it is stored in a **.env** file that is not tracked by version control and if not configured in a **.env**, all sessions and other encrypted data will not be secure.

File: contracts/auth.ts (hunk 8 above)
Comment (arnoldkunihira) [reply contracts/auth.ts:10]: The most obvious benefit is, adonisjs lets me focus on building the application, so I don't have to focus much on building architecture around it as it comes with DB setup, mails, MVC, testing utilities, social authentication, you name it, and it is also well structured compared to expressjs and others.

File: config/auth.ts (hunk 5 above)
Comment (arnoldkunihira) [reply config/auth.ts:36]: An OAT has a format that is not intended to be read by anyone hence the word opaque in OAT, it is only the issuer that knows the format of this token while a JWT has readable content as anyone can decode the token and read the information in it.

File: app/Models/User.ts (hunk 4 above)
Comment (arnoldkunihira) [reply app/Models/User.ts:27]: The best key takeaway is it has evolved from versions 3 and 4 where some things were missing and development was rather complex using this framework, and now it is on course to compete with widely used frameworks like express, and from a development point, it makes development faster as it consists mostly of the packages one needs to get an application up and running in not time.

File: .env.example (hunk 1 above)
Comment (vongohren) [reply .env.example:4]: Okok, the offer encryption features out of the box. I see. But you did check in your app key :P I guess that was a mistake? @arnoldkunihira

File: app/Models/User.ts (hunk 4 above)
Comment (vongohren) [reply app/Models/User.ts:27]: Sounds good!

**Pull request merged**